### PR TITLE
New resource duplocloud_plan_settings added.

### PR DESCRIPTION
**Pull request by tahir-duplo**
DUPLO-2959: New resource duplocloud_plan_settings added.

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-2959 Need a duplocloud_plan_settings resource](https://app.clickup.com/t/8655600/DUPLO-2959)

File: duplocloud/resource_duplo_lbconfigs.go [hunk 1]
```diff
@@ -396,7 +396,7 @@ func duploServiceLbConfigsWaitUntilReady(ctx context.Context, c *duplosdk.Client
 			// Find a cloud load balancer, and get it's status.
 			isCloudLb := false
 			for _, lb := range *list {
-				if lb.LbType != 2 && lb.LbType != 3 && lb.LbType != 4 {
+				if lb.LbType != 0 && lb.LbType != 2 && lb.LbType != 3 && lb.LbType != 4 {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_lbconfigs.go:399]: Can we keep this out of this PR?  It is already on another PR, that I commented on.

File: duplocloud/resource_duplo_plan_settings.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+func resourcePlanSettingsDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	planID := d.Get("plan_id").(string)
+	log.Printf("[TRACE] resourcePlanSettingsDelete(%s): start", planID)
+
+	c := m.(*duplosdk.Client)
+
+	duplo, err := c.PlanGet(planID)
+	if err != nil && !err.PossibleMissingAPI() {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_plan_settings.go:134]: We should not check for possible missing API here, just let it an error.  Possible missing API checks are meant to "fallback" from newer APIs to older APIs - and this is an old API with no alternative.

File: duplocloud/resource_duplo_plan_settings.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+func resourcePlanSettingsRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	planID := d.Id()
+	log.Printf("[TRACE] resourcePlanSettingsRead(%s): start", planID)
+
+	c := m.(*duplosdk.Client)
+
+	duplo, err := c.PlanGet(planID)
+	if err != nil && !err.PossibleMissingAPI() {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_plan_settings.go:80]: We should not check for possible missing API here, just let it an error.  Possible missing API checks are meant to "fallback" from newer APIs to older APIs - and this is an old API with no alternative.

File: duplocloud/resource_duplo_plan_settings.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	planID := d.Get("plan_id").(string)
+	log.Printf("[TRACE] resourcePlanSettingsDelete(%s): start", planID)
+
+	c := m.(*duplosdk.Client)
+
+	duplo, err := c.PlanGet(planID)
+	if err != nil && !err.PossibleMissingAPI() {
+		return diag.Errorf("failed to retrieve plan for '%s': %s", planID, err)
+	}
+	if _, ok := d.GetOk("unrestricted_ext_lb"); ok {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_plan_settings.go:137]: We should check for a nil `duplo` result from the API call. If nil, skip the delete operation.

File: duplocloud/resource_duplo_plan_settings.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	planID := d.Get("plan_id").(string)
+	log.Printf("[TRACE] resourcePlanSettingsCreateOrUpdate(%s): start", planID)
+
+	c := m.(*duplosdk.Client)
+
+	duplo, err := c.PlanGet(planID)
+	if err != nil && !err.PossibleMissingAPI() {
+		return diag.Errorf("failed to retrieve plan for '%s': %s", planID, err)
+	}
+	if v, ok := d.GetOk("unrestricted_ext_lb"); ok {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_plan_settings.go:110]: We should check for a nil `duplo` result from the API call. If nil, the provider should return an error that the plan could not be found.

File: duplocloud/resource_duplo_plan_settings.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+func resourcePlanSettingsCreateOrUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	planID := d.Get("plan_id").(string)
+	log.Printf("[TRACE] resourcePlanSettingsCreateOrUpdate(%s): start", planID)
+
+	c := m.(*duplosdk.Client)
+
+	duplo, err := c.PlanGet(planID)
+	if err != nil && !err.PossibleMissingAPI() {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_plan_settings.go:107]: We should not check for possible missing API here, just let it an error.  Possible missing API checks are meant to "fallback" from newer APIs to older APIs - and this is an old API with no alternative.

**Review by joek-duplo — CHANGES_REQUESTED** [review]

File: duplocloud/resource_duplo_lbconfigs.go (hunk 1 above)
Comment (tahir-duplo) [reply duplocloud/resource_duplo_lbconfigs.go:399]: Done.

File: duplocloud/resource_duplo_plan_settings.go (hunk 2 above)
Comment (tahir-duplo) [reply duplocloud/resource_duplo_plan_settings.go:134]: Done.

File: duplocloud/resource_duplo_plan_settings.go (hunk 3 above)
Comment (tahir-duplo) [reply duplocloud/resource_duplo_plan_settings.go:80]: Done.

File: duplocloud/resource_duplo_plan_settings.go (hunk 5 above)
Comment (tahir-duplo) [reply duplocloud/resource_duplo_plan_settings.go:110]: Done.

File: duplocloud/resource_duplo_plan_settings.go (hunk 6 above)
Comment (tahir-duplo) [reply duplocloud/resource_duplo_plan_settings.go:107]: Done.

File: duplocloud/resource_duplo_plan_settings.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func resourcePlanSettingsCreateOrUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	planID := d.Get("plan_id").(string)
+	log.Printf("[TRACE] resourcePlanSettingsCreateOrUpdate(%s): start", planID)
+
+	c := m.(*duplosdk.Client)
+
+	duplo, err := c.PlanGet(planID)
+	if duplo == nil {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_plan_settings.go:99]: The error check must come before the `duplo` nil check, or we end up not reporting errors to the user.

File: duplocloud/resource_duplo_plan_settings.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+func resourcePlanSettingsDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	planID := d.Get("plan_id").(string)
+	log.Printf("[TRACE] resourcePlanSettingsDelete(%s): start", planID)
+
+	c := m.(*duplosdk.Client)
+
+	duplo, err := c.PlanGet(planID)
+	// Skip if plan does not exist.
+	if duplo == nil {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_plan_settings.go:130]: The error check must come before the `duplo` nil check, or we end up not reporting errors to the user.

File: duplocloud/resource_duplo_plan_settings.go [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func resourcePlanSettingsRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	planID := d.Id()
+	log.Printf("[TRACE] resourcePlanSettingsRead(%s): start", planID)
+
+	c := m.(*duplosdk.Client)
+
+	duplo, err := c.PlanGet(planID)
+	if duplo == nil {
```
Comment (joek-duplo) [inline duplocloud/resource_duplo_plan_settings.go:80]: The error check must come before the `duplo` nil check, or we end up not reporting errors to the user.

**Review by joek-duplo — CHANGES_REQUESTED** [review]

**Review by joek-duplo — APPROVED** [review]

**Review by joek-duplo — APPROVED** [review]

**Review by joek-duplo — APPROVED** [review]

**Pull request merged**